### PR TITLE
Napps 1591: Split dockerfiles for rake command

### PR DESCRIPTION
**Pull request by aaronbrezel**
This PR adds a new dockerfile dedicated to the `rake check:all` command. For clarity, the original Klaxon server `Dockerfile` has been deleted--with `Dockerfile.server` and `Dockerfile.rake` replacing it. Both Dockerfiles will employ the exact same build setup. 

The primary difference is the entrypoint command. For `Dockerfile.server` we use the same entrypoint as the original `Dockerfile`. The `Dockerfile.rake` image will execute `rake check:all` instead. 

The image created by `Dockerfile.rake` will end up being used by the ECS scheduled task described in [NAPPS-1591]. 

The only other change of note, is the optional build arg in `Dockerfile.server` which allows local developers to continue using the cron job to run `rake check:all` despite it being phased out of production. 

## To test

Run 
```
docker-compose up --build --force-recreate
```

We should see three bits of behavior on startup 
1. Three containers should be built: `klaxon-db-1`, `klaxon-app-1` and `klaxon-rake-1`. 
2. `klaxon-rake-1` should execute a `rake check:all` command (with email style output) and terminal safely
3. `klaxon-db-1` and `klaxon-app-1` should continue as they always have, creating a working Klaxon server. 
4. After 10 minutes, you should see `klaxon-app-1` execute `rake check:all` through its cron timer. 

[NAPPS-1591]: https://arcpublishing.atlassian.net/browse/NAPPS-1591?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: app/views/sessions/unknown_user.html.erb [hunk 1]
```diff
@@ -11,7 +11,7 @@
     </div>
     <div class="row">
       <div class="col-md-6">
-      <p class="lead">If you have an existing account, you might have typed your email wrong. <a href="/">Try again.</a> If not, contact your adminstrator.</p>
+      <p class="lead">If you have an existing account, you might have typed your email wrong. <a href="/klaxon">Try again.</a> If not, contact your adminstrator.</p>
```
Comment (aaronbrezel) [inline app/views/sessions/unknown_user.html.erb:14]: Just an unrelated little fix for this broken link

File: docker-entrypoint.sh [hunk 2]
```diff
@@ -1,8 +1,13 @@
...
-echo "Calling script to set up server cronjobs"
-/bin/bash scripts/cron_setup.sh
+if [[ "$USE_CRON" == "true" ]]; then
+    # Add crontab jobs
+    echo "Calling script to set up server cronjobs"
+    /bin/bash scripts/cron_setup.sh
+else
+    echo "Skipping cron setup"
+fi
+
```
Comment (aaronbrezel) [inline docker-entrypoint.sh]: Mechanism to allow local development to use the cron to execute `rake check:all`

File: Dockerfile.server [hunk 3]
```diff
@@ -1,7 +1,7 @@
 FROM public.ecr.aws/docker/library/ruby:2.7.7
 
-# Throw errors if Gemfile has been modified since Gemfile.lock
-RUN bundle config --global frozen 1
+ARG USE_CRON=false
```
Comment (aaronbrezel) [inline Dockerfile.server:3]: This build arg is used to determine whether Klaxon server container should setup the `rake check:all` command on a cron timer. Since the default value is fault, we don't need to worry about setting this variable during ECS deployment

File: app/views/sessions/unknown_user.html.erb (hunk 1 above)
Comment (emlorraine) [reply app/views/sessions/unknown_user.html.erb:14]: Good catch!

**Comment by emlorraine** [comment]
Thanks for working through this! I ran the testing instructions and verify that the three containers are present. I'm not seeing `rake check:all` execution in my terminal history or email-style output. Are there other permissions/testing steps we should configure to get that output?

**Review by emlorraine — CHANGES_REQUESTED** [review]
Thanks for working through this, @aaronbrezel! I ran the testing instructions and verify that the three containers are present. The echo `Calling script to set up server cronjobs` is in my terminal, but I'm not seeing `rake check:all` execution in my terminal history or email-style output. Are there other permissions/testing steps we should configure to get that output?

File: app/views/sessions/unknown_user.html.erb (hunk 1 above)
Comment (Kat-Alo) [reply app/views/sessions/unknown_user.html.erb:14]: I think this was maybe ticketed out 😂 but happy to include

File: buildspec.yml [hunk 4]
```diff
@@ -23,8 +23,13 @@ phases:
 
   build:
     on-failure: ABORT
-    commands:
-      - docker build --cache-from $AWS_ACCOUNT_ID.dkr.ecr.$AWS_DEFAULT_REGION.amazonaws.com/$APP_NAME:$DEPLOYMENT_ENV-latest -t $APP_NAME:$DEPLOYMENT_ENV-`cat commit_hash` -f $DOCKERFILE_PATH .
+    commands: # Temporarily set USE_CRON=true while we debug ECS scheduled task
+      - | 
+        docker build \
+          --cache-from $AWS_ACCOUNT_ID.dkr.ecr.$AWS_DEFAULT_REGION.amazonaws.com/$APP_NAME:$DEPLOYMENT_ENV-latest \
+          --build-arg USE_CRON=true \
```
Comment (Kat-Alo) [inline buildspec.yml:30]: 💇‍♀️

**Review by Kat-Alo — COMMENTED** [review]
:)

File: docker-compose.yml [hunk 5]
```diff
@@ -15,6 +15,18 @@ services:
...
       - .env.local
-    build: .
+    build:
+      context: .
+      dockerfile: Dockerfile.server
+      args:
+        - USE_CRON=true
+    depends_on:
+      - db
+  rake:
```
Comment (Kat-Alo) [inline docker-compose.yml]: Recapping our zoom! Plan B: if we can't get the initial invocation of the `rake check:all` command in `docker-entrypoint.`🤐  or into the cronjob itself, then let's plop an inline comment here explaining that this `rake` container really only serves to execute the first command for developer experience, all subsequent calls are via the cronjob

File: docker-entrypoint.sh [hunk 6]
```diff
@@ -1,8 +1,13 @@
 #!/bin/bash
 
-# Add crontab jobs
-echo "Calling script to set up server cronjobs"
-/bin/bash scripts/cron_setup.sh
+if [[ "$USE_CRON" == "true" ]]; then
+    # Add crontab jobs
+    echo "Calling script to set up server cronjobs"
+    /bin/bash scripts/cron_setup.sh
```
Comment (Kat-Alo) [inline docker-entrypoint.sh:6]: Recapping our zoom continued! Plan A: hoping we can get the first call to `rake check:all` to run either in the cron job or here. That way, there's a very clear distinction that the page checks happen in cron locally and in the rake task in the deployed environment.

Comment (aaronbrezel) [reply docker-entrypoint.sh:6]: Play A was a success with `bundle exec rake check:all` added to the docker_entrypoint

File: docker-entrypoint.sh [hunk 7]
```diff
@@ -8,6 +8,9 @@ else
     echo "Skipping cron setup"
 fi
 
+# Initialize app with initial diff check on watched pages
+echo "Initial check on watched pages" 
+bundle exec rake check:all
```
Comment (Kat-Alo) [inline docker-entrypoint.sh]: question! would this be better suited for the `if [[ "$USE_CRON" == "true" ]]; then` block? That way, once the ECS task scheduler is set up and we remove the line setting `USE_CRON` to `true`, we won't have to relocate this? Because I imagine similarly to how it's slightly preferable to only have local `rake check:all` calls made in the cronjob (or close to it), it'd also be preferable to only run the page checks with the ecs scheduler in the deployed env.

**Review by Kat-Alo — COMMENTED** [review]
just the one thought! HAPPY WEEKEND!

File: docker-entrypoint.sh (hunk 7 above)
Comment (aaronbrezel) [reply docker-entrypoint.sh]: Ahh, you're totally right. When I initially put this together, I was thinking that we're run this initial `rake check:all` in both environments. But you remind me that we only really need to do this in the local env.

**Review by Kat-Alo — APPROVED** [review]

**Review by emlorraine — APPROVED** [review]
Thank you for taking on this work! I was able to verify the containers build and generate some email output ✔️
